### PR TITLE
Fix textEdit

### LINKED ISSUE #483: `clangd` format issues
I found that the formatting function with `clangd` is sometimes a hit or miss. I have the following example to reproduce the issue that I have seen.

Create the following file with the given content:
```cpp
class ABC {
    private:
int a;};
```
Then running `LspDocumentFormat` gives me the following result:
```cpp
class ABC {private:

int a;};
```

The expected formatted document should be as follows:
```cpp
class ABC {
private:
  int a;
};
```

This is when no `.clang-format` file is used.

The log file is attached here: 
[vim-lsp.log](https://github.com/prabirshrestha/vim-lsp/files/3563740/vim-lsp.log)

### PR DESCRIPTION
**Pull request by mattn**
The p comand append empty line. So when replacing text block with "\n", it need
to remove with dd.

Fixes #483

### CODE REVIEW

File: autoload/lsp/utils/text_edit.vim [hunk 1]
```diff
@@ -202,6 +202,9 @@ function! s:generate_sub_cmd_replace(text_edit) abort
     else
         let l:sub_cmd .= "\"=l:merged_text_edit['merged']['newText']\<CR>p"
     endif
+    if l:new_text =~# '^\n$'
+        let l:sub_cmd .= 'dd'
```
Comment (prabirshrestha) [inline autoload/lsp/utils/text_edit.vim]: If we use `dd` would this change the register/clipboard contents?

Comment (thomasfaingnaert) [reply autoload/lsp/utils/text_edit.vim]: I think it will. This might be better?

```suggestion
        let l:sub_cmd .= '"_dd'
```

**Comment by prabirshrestha** [comment]
Can we have test cases for textedit changes? https://github.com/prabirshrestha/vim-lsp/blob/master/test/lsp/utils/text_edit.vimspec

Also wondering if instead of using `_dd` we just remove the last char. Wondering if that would solve the issue mentioned here https://github.com/prabirshrestha/vim-lsp/issues/483#issuecomment-529135944

**Comment by mattn** [comment]
I noticed current implementation is hard to fix this issue. Expression register can not insert text with "\n".

```
new
call setline('.', ['xxxxyyyy'])
let s:foo = "foo\nbar\n"
execute "normal! 1G4l\"=s:foo\<CR>P"
```
We expect this result is

```
xxxxfoo
baryyyy
```
But real result is
```
foo
bar
xxxx
yyyy
```
This is written in `:help quote=`.

**Comment by mattn** [comment]
I tried to fix this issue with using `i_CTRL-R=foo` but this way break `v:completed_item` since this goes insert mode temporary.

**Comment by thomasfaingnaert** [comment]
@mattn 
It works if you `let s:foo = "foo\nbar"` (i.e. remove the newline at the end).
I'm not sure how we would handle cases where you do want the newline, though.

**Comment by thomasfaingnaert** [comment]
@mattn
How about something like this?
```vim
new
call setline('.', ['xxxxyyyy'])
let s:foo = "foo\nbar\n" . '?'
execute "normal! 1G4l\"=s:foo\<CR>gPh\"_x"
```

This adds a `?` and later deletes it using `h"_x`. This way, we don't have to worry about newlines.

**Comment by mattn** [comment]
Fixed.

File: autoload/lsp/utils/text_edit.vim [hunk 2]
```diff
@@ -162,15 +162,14 @@ endfunction
 function! s:generate_sub_cmd_insert(text_edit) abort
     let l:start_line = a:text_edit['range']['start']['line']
     let l:start_character = a:text_edit['range']['start']['character']
-    let l:new_text = s:parse(a:text_edit['newText'])
```
Comment (thomasfaingnaert) [inline autoload/lsp/utils/text_edit.vim]: We can remove `s:parse` too now, right?

Comment (mattn) [reply autoload/lsp/utils/text_edit.vim]: removed

**Comment by mattn** [comment]
Thanks you all!

**Comment by thomasfaingnaert** [comment]
@mattn You forgot to squash again :p

**Comment by mattn** [comment]
Oh... my fault.

**Comment by thenewvu** [comment]
The issue still happens to me, please try similar steps below:
```
svg_stroke(|)
svg_stroke|()
svg_|()
choose one of completion item, ex: fill, then <esc>:
svg_fill
-> notice the () has been removed.

**Comment by thomasfaingnaert** [comment]
@mattn If you go to a PR (say, https://github.com/prabirshrestha/vim-lsp/pull/464) and click the dropdown to select "squash and merge", you can set it as default.

**Comment by thomasfaingnaert** [comment]
@thenewvu Do you mean #482?

**Comment by thenewvu** [comment]
yeah ... oh .. wait, I thought this PR fixes the issue ?

**Comment by thomasfaingnaert** [comment]
I haven't looked into it yet, but it seems to be related to asyncomplete?

**Comment by thenewvu** [comment]
Not sure, but `let g:lsp_text_edit_enabled = 0` workarounds the issue, it relates to https://github.com/prabirshrestha/asyncomplete.vim/issues/156

**Pull request merged**